### PR TITLE
Gateway Changes

### PR DESCRIPTION
**Pull request by mappum**
Miscellaneous changes and fixes for the gateway and command HTTP handler.

* Security fixes (Suborigins, CSRF protection, API server path whitelist)
* Turn on gateway server by default in initial config
* Fixed multipart filename escape bug (now supports names with quotes)

### CODE REVIEW

**Comment by jbenet** [comment]
This is getting rebased on top of @briantigerchow's branch, right?

**Comment by briantigerchow** [comment]
@mappum let's rebase on #741 and iron out the changes

**Comment by briantigerchow** [comment]
awaiting LGTM from @jbenet

File: commands/http/handler.go [hunk 1]
```diff
@@ -55,6 +56,20 @@ func (i Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
...
 
+	// error on external referers (to prevent CSRF attacks)
+	referer := r.Referer()
+	scheme := r.URL.Scheme
+	if len(scheme) == 0 {
+		scheme = "http"
+	}
+	host := fmt.Sprintf("%s://%s/", scheme, r.Host)
+	// empty string means the user isn't following a link (they are directly typing in the url)
+	if referer != "" && !strings.HasPrefix(referer, host) {
```
Comment (jbenet) [inline commands/http/handler.go]: Hmm, wont this break sharness, too? or curling? oh i guess those dont set the referrer? The security of the API will probably have to move towards tokens or something in the future.

File: core/corehttp/gateway_handler.go [hunk 2]
```diff
@@ -193,6 +194,12 @@ func (i *gatewayHandler) getHandler(w http.ResponseWriter, r *http.Request) {
 
 	w.Header().Set("X-IPFS-Path", p)
 
+	// Suborigin header, sandboxes apps from each other in the browser (even
+	// though they are served from the same gateway domain). NOTE: This is not
+	// yet widely supported by browsers.
```
Comment (jbenet) [inline core/corehttp/gateway_handler.go]: :+1: ^ :100:

**Comment by jbenet** [comment]
LGTM!

**Comment by jbenet** [comment]
hmm jenkins didnt run?

**Comment by mappum** [comment]
It's green :+1:

**Comment by jbenet** [comment]
:heart: :+1:

**Pull request merged**